# Loading the dataset

In [1]:
import pandas as pd

df_peugot_1 = pd.read_csv("C:/Users/hp/Desktop/jagdish/archive/peugeot_207_01.csv", delimiter=';')
df_peugot_2 = pd.read_csv("C:/Users/hp/Desktop/jagdish/archive/peugeot_207_02.csv", delimiter=';')

df_corsa_1 = pd.read_csv("C:/Users/hp/Desktop/jagdish/archive/opel_corsa_01.csv", delimiter=';')
df_corsa_2 = pd.read_csv("C:/Users/hp/Desktop/jagdish/archive/opel_corsa_02.csv", delimiter=';')



df_new = pd.DataFrame(columns=list(df_corsa_1.columns))
print(df_new.head(5))

df_new = pd.concat([df_peugot_1, df_peugot_2, df_corsa_1, df_corsa_2], axis=0)
print(df_new.shape)

Empty DataFrame
Columns: [AltitudeVariation, VehicleSpeedInstantaneous, VehicleSpeedAverage, VehicleSpeedVariance, VehicleSpeedVariation, LongitudinalAcceleration, EngineLoad, EngineCoolantTemperature, ManifoldAbsolutePressure, EngineRPM, MassAirFlow, IntakeAirTemperature, VerticalAcceleration, FuelConsumptionAverage, roadSurface, traffic, drivingStyle]
Index: []
(24957, 17)


In [2]:
print(df_peugot_1.shape)
print(df_peugot_2.shape)
print(df_corsa_1.shape)
print(df_corsa_2.shape)

(8614, 17)
(4623, 17)
(7392, 17)
(4328, 17)


In [3]:
df_new.head(10)

AltitudeVariation VehicleSpeedInstantaneous VehicleSpeedAverage  \
0               NaN                         0                 NaN   
1               NaN                         0                 NaN   
2               NaN                         0                 NaN   
3               NaN                         0                 NaN   
4               NaN                         0                 NaN   
5               NaN                         0                 NaN   
6               NaN                         0                 NaN   
7               NaN                         0                 NaN   
8               NaN                         0                 NaN   
9      -2,800003052                         0                 NaN   

  VehicleSpeedVariance VehicleSpeedVariation LongitudinalAcceleration  \
0                  NaN                   NaN                   0,0418   
1                  NaN                   NaN                   0,0713   
2                  NaN                   NaN                   0,0952   
3                  NaN                   NaN                   0,1136   
4                  NaN                   NaN                   0,1274   
5                  NaN                   NaN                   0,1396   
6                  NaN                   NaN                   0,1589   
7                  NaN                   NaN                   0,1951   
8                  NaN                   NaN                   0,2408   
9                  NaN                   NaN                    0,279   

    EngineLoad  EngineCoolantTemperature  ManifoldAbsolutePressure EngineRPM  \
0  36,47058868                      28.0                     100.0     802,5   
1  36,47058868                      28.0                     100.0     802,5   
2  36,47058868                      29.0                     100.0     798,5   
3  36,47058868                      29.0                     100.0     798,5   
4  36,47058868                      29.0                     100.0     804,5   
5  36,47058868                      29.0                     100.0     804,5   
6  36,47058868                      29.0                     100.0     800,5   
7  36,07843018                      29.0                     100.0       800   
8  36,07843018                      29.0                     100.0       800   
9  36,47058868                      29.0                     100.0       782   

   MassAirFlow  IntakeAirTemperature VerticalAcceleration  \
0  10,30000019                  14.0               0,0118   
1  10,30000019                  14.0              -0,0023   
2  10,23999977                  14.0              -0,0135   
3  10,23999977                  15.0              -0,0161   
4  10,35999966                  15.0               -0,016   
5  10,35999966                  15.0              -0,0147   
6  10,32999992                  15.0              -0,0108   
7  10,32999992                  15.0              -0,0209   
8  10,40999985                  15.0              -0,0446   
9  10,22000027                  15.0              -0,0571   

  FuelConsumptionAverage      roadSurface                 traffic  \
0                    NaN  SmoothCondition  LowCongestionCondition   
1                    NaN  SmoothCondition  LowCongestionCondition   
2                    NaN  SmoothCondition  LowCongestionCondition   
3                    NaN  SmoothCondition  LowCongestionCondition   
4                    NaN  SmoothCondition  LowCongestionCondition   
5                    NaN  SmoothCondition  LowCongestionCondition   
6                    NaN  SmoothCondition  LowCongestionCondition   
7                    NaN  SmoothCondition  LowCongestionCondition   
8                    NaN  SmoothCondition  LowCongestionCondition   
9                    NaN  SmoothCondition  LowCongestionCondition   

    drivingStyle  
0  EvenPaceStyle  
1  EvenPaceStyle  
2  EvenPaceStyle  
3  EvenPaceStyle  
4  EvenPaceStyle  
5  EvenPace

In [4]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24957 entries, 0 to 4327
Data columns (total 17 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   AltitudeVariation          24777 non-null  object 
 1   VehicleSpeedInstantaneous  24913 non-null  object 
 2   VehicleSpeedAverage        23775 non-null  object 
 3   VehicleSpeedVariance       23775 non-null  object 
 4   VehicleSpeedVariation      24769 non-null  object 
 5   LongitudinalAcceleration   24957 non-null  object 
 6   EngineLoad                 24952 non-null  object 
 7   EngineCoolantTemperature   24952 non-null  float64
 8   ManifoldAbsolutePressure   24952 non-null  float64
 9   EngineRPM                  24952 non-null  object 
 10  MassAirFlow                24952 non-null  object 
 11  IntakeAirTemperature       24952 non-null  float64
 12  VerticalAcceleration       24957 non-null  object 
 13  FuelConsumptionAverage     24671 non-null  obje

In [7]:
df_new.describe()
df_new = df_new.fillna(0)

In [8]:
def process_col_values(df_new, col_name):
    col_list = list(df_new[col_name])
    processed_value = []
    for each_value in col_list:
        if each_value == 0:
            processed_value.append(0)
        elif type(each_value) == str:
            processed_value.append(float(each_value.replace(",",".")))
        else:
            processed_value.append(float(each_value))
    
    df_new[col_name] = processed_value

In [9]:
cols_to_change = ["AltitudeVariation", "VehicleSpeedInstantaneous", "VehicleSpeedAverage", "VehicleSpeedVariance", "VehicleSpeedVariation",
                 "LongitudinalAcceleration", "EngineLoad", "EngineRPM", "MassAirFlow", "VerticalAcceleration", "FuelConsumptionAverage"]

for col in cols_to_change:
    print(col)
    process_col_values(df_new, col)

AltitudeVariation
VehicleSpeedInstantaneous
VehicleSpeedAverage
VehicleSpeedVariance
VehicleSpeedVariation
LongitudinalAcceleration
EngineLoad
EngineRPM
MassAirFlow
VerticalAcceleration
FuelConsumptionAverage


In [10]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24957 entries, 0 to 4327
Data columns (total 17 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   AltitudeVariation          24957 non-null  float64
 1   VehicleSpeedInstantaneous  24957 non-null  float64
 2   VehicleSpeedAverage        24957 non-null  float64
 3   VehicleSpeedVariance       24957 non-null  float64
 4   VehicleSpeedVariation      24957 non-null  float64
 5   LongitudinalAcceleration   24957 non-null  float64
 6   EngineLoad                 24957 non-null  float64
 7   EngineCoolantTemperature   24957 non-null  float64
 8   ManifoldAbsolutePressure   24957 non-null  float64
 9   EngineRPM                  24957 non-null  float64
 10  MassAirFlow                24957 non-null  float64
 11  IntakeAirTemperature       24957 non-null  float64
 12  VerticalAcceleration       24957 non-null  float64
 13  FuelConsumptionAverage     24957 non-null  floa

In [11]:
df_new.tail()

AltitudeVariation  VehicleSpeedInstantaneous  VehicleSpeedAverage  \
4323          -2.800003                  64.176300            57.469687   
4324          -2.800003                  63.275578            57.616054   
4325          -2.800003                  62.600033            57.754915   
4326          -2.800003                  62.600033            57.859999   
4327          -2.800003                  61.474136            57.965083   

      VehicleSpeedVariance  VehicleSpeedVariation  LongitudinalAcceleration  \
4323             76.346241              -0.900719                     0.305   
4324             76.745659              -0.900722                     0.383   
4325             76.957120              -0.675545                     0.437   
4326             77.307649               0.000000                     0.496   
4327             77.395081              -1.125896                     0.543   

      EngineLoad  EngineCoolantTemperature  ManifoldAbsolutePressure  \
4323   17.254902                      83.0                     108.0   
4324   20.784313                      83.0                     109.0   
4325   18.823530                      82.0                     108.0   
4326   11.764706                      82.0                     106.0   
4327    3.529412                      82.0                     105.0   

      EngineRPM  MassAirFlow  IntakeAirTemperature  VerticalAcceleration  \
4323     2009.0        14.36                  18.0               -0.0039   
4324     1996.0        13.83                  18.0                0.0234   
4325     1982.0        13.81                  18.0                0.0469   
4326     1952.0        13.38                  18.0                0.0586   
4327     1924.0        12.51                  18.0                0.0273   

      FuelConsumptionAverage      roadSurface                 traffic  \
4323                8.389318  SmoothCondition  LowCongestionCondition   
4324                8.357814  SmoothCondition  LowCongestionCondition   
4325                8.332450  SmoothCondition  LowCongestionCondition   
4326                8.305367  SmoothCondition  LowCongestionCondition   
4327                8.275750  SmoothCondition  LowCongestionCondition   

       drivingStyle  
4323  EvenPaceStyle  
4324  EvenPaceStyle  
4325  EvenPaceStyle  
4326  EvenPaceStyle  
4327  EvenPaceStyle

In [12]:
road_surf_unique = list(df_new["roadSurface"].unique())
traffic_unique = list(df_new["traffic"].unique())
drivingStyle_unique = list(df_new["drivingStyle"].unique())

print(road_surf_unique)
print(traffic_unique)
print(drivingStyle_unique)

['SmoothCondition', 'FullOfHolesCondition', 'UnevenCondition']
['LowCongestionCondition', 'NormalCongestionCondition', 'HighCongestionCondition']
['EvenPaceStyle', 'AggressiveStyle']


In [13]:
final_list = []
final_list.extend(road_surf_unique)
final_list.extend(road_surf_unique)
final_list.extend(road_surf_unique)
print(final_list)

['SmoothCondition', 'FullOfHolesCondition', 'UnevenCondition', 'SmoothCondition', 'FullOfHolesCondition', 'UnevenCondition', 'SmoothCondition', 'FullOfHolesCondition', 'UnevenCondition']


In [14]:
one_hot_encoding = pd.get_dummies(df_new[['roadSurface', "traffic", "drivingStyle"]])
one_hot_encoding

roadSurface_FullOfHolesCondition  roadSurface_SmoothCondition  \
0                                    0                            1   
1                                    0                            1   
2                                    0                            1   
3                                    0                            1   
4                                    0                            1   
...                                ...                          ...   
4323                                 0                            1   
4324                                 0                            1   
4325                                 0                            1   
4326                                 0                            1   
4327                                 0                            1   

      roadSurface_UnevenCondition  traffic_HighCongestionCondition  \
0                               0                                0   
1                               0                                0   
2                               0                                0   
3                               0                                0   
4                               0                                0   
...                           ...                              ...   
4323                            0                                0   
4324                            0                                0   
4325                            0                                0   
4326                            0                                0   
4327                            0                                0   

      traffic_LowCongestionCondition  traffic_NormalCongestionCondition  \
0                                  1                                  0   
1                                  1                                  0   
2                                  1                                  0   
3                                  1                                  0   
4                                  1                                  0   
...                              ...                                ...   
4323                               1                                  0   
4324                               1                                  0   
4325                               1                                  0   
4326                               1                                  0   
4327                               1                                  0   

      drivingStyle_AggressiveStyle  drivingStyle_EvenPaceStyle  
0                                0                           1  
1                                0                           1  
2                                0                           1  
3                                0                           1  
4                                0                           1  
...                            ...                         ...  
4323                             0                           1  
4324                             0                           1  
4325                             0                           1  
4326                             0                           1  
4327                             0                           1  

[24957 rows x 8 columns]

In [15]:
one_hot_encoding['Target']= one_hot_encoding.values.tolist()
one_hot_encoding

roadSurface_FullOfHolesCondition  roadSurface_SmoothCondition  \
0                                    0                            1   
1                                    0                            1   
2                                    0                            1   
3                                    0                            1   
4                                    0                            1   
...                                ...                          ...   
4323                                 0                            1   
4324                                 0                            1   
4325                                 0                            1   
4326                                 0                            1   
4327                                 0                            1   

      roadSurface_UnevenCondition  traffic_HighCongestionCondition  \
0                               0                                0   
1                               0                                0   
2                               0                                0   
3                               0                                0   
4                               0                                0   
...                           ...                              ...   
4323                            0                                0   
4324                            0                                0   
4325                            0                                0   
4326                            0                                0   
4327                            0                                0   

      traffic_LowCongestionCondition  traffic_NormalCongestionCondition  \
0                                  1                                  0   
1                                  1                                  0   
2                                  1                                  0   
3                                  1                                  0   
4                                  1                                  0   
...                              ...                                ...   
4323                               1                                  0   
4324                               1                                  0   
4325                               1                                  0   
4326                               1                                  0   
4327                               1                                  0   

      drivingStyle_AggressiveStyle  drivingStyle_EvenPaceStyle  \
0                                0                           1   
1                                0                           1   
2                                0                           1   
3                                0                           1   
4                                0                           1   
...                            ...                         ...   
4323                             0                           1   
4324                             0                           1   
4325                             0                           1   
4326                             0                           1   
4327                             0                           1   

                        Target  
0     [0, 1, 0, 0, 1, 0, 0, 1]  
1     [0, 1, 0, 0, 1, 0, 0, 1]  
2     [0, 1, 0, 0, 1, 0, 0, 1]  
3     [0, 1, 0, 0, 1, 0, 0, 1]  
4     [0, 1, 0, 0, 1, 0, 0, 1]  
...                        ...  
4323  [0, 1, 0, 0, 1, 0, 0, 1]  
4324  [0, 1, 0, 0, 1, 0, 0, 1]  
4325  [0, 1, 0, 0, 1, 0, 0, 1]  
4326  [0, 1, 0, 0, 1, 0, 0, 1]  
4327  [0, 1, 0, 0, 1, 0, 0, 1]  

[24957 rows x 9 columns]

In [16]:
df_processed = df_new.drop(["roadSurface", "traffic", "drivingStyle"], axis = 1)
df_processed['Target'] = one_hot_encoding['Target']
df_processed

AltitudeVariation  VehicleSpeedInstantaneous  VehicleSpeedAverage  \
0              0.000000                   0.000000             0.000000   
1              0.000000                   0.000000             0.000000   
2              0.000000                   0.000000             0.000000   
3              0.000000                   0.000000             0.000000   
4              0.000000                   0.000000             0.000000   
...                 ...                        ...                  ...   
4323          -2.800003                  64.176300            57.469687   
4324          -2.800003                  63.275578            57.616054   
4325          -2.800003                  62.600033            57.754915   
4326          -2.800003                  62.600033            57.859999   
4327          -2.800003                  61.474136            57.965083   

      VehicleSpeedVariance  VehicleSpeedVariation  LongitudinalAcceleration  \
0                 0.000000               0.000000                    0.0418   
1                 0.000000               0.000000                    0.0713   
2                 0.000000               0.000000                    0.0952   
3                 0.000000               0.000000                    0.1136   
4                 0.000000               0.000000                    0.1274   
...                    ...                    ...                       ...   
4323             76.346241              -0.900719                    0.3050   
4324             76.745659              -0.900722                    0.3830   
4325             76.957120              -0.675545                    0.4370   
4326             77.307649               0.000000                    0.4960   
4327             77.395081              -1.125896                    0.5430   

      EngineLoad  EngineCoolantTemperature  ManifoldAbsolutePressure  \
0      36.470589                      28.0                     100.0   
1      36.470589                      28.0                     100.0   
2      36.470589                      29.0                     100.0   
3      36.470589                      29.0                     100.0   
4      36.470589                      29.0                     100.0   
...          ...                       ...                       ...   
4323   17.254902                      83.0                     108.0   
4324   20.784313                      83.0                     109.0   
4325   18.823530                      82.0                     108.0   
4326   11.764706                      82.0                     106.0   
4327    3.529412                      82.0                     105.0   

      EngineRPM  MassAirFlow  IntakeAirTemperature  VerticalAcceleration  \
0         802.5        10.30                  14.0                0.0118   
1         802.5        10.30                  14.0               -0.0023   
2         798.5        10.24                  14.0               -0.0135   
3         798.5        10.24                  15.0               -0.0161   
4         804.5        10.36                  15.0               -0.0160   
...         ...          ...                   ...                   ...   
4323     2009.0        14.36                  18.0               -0.0039   
4324     1996.0        13.83                  18.0                0.0234   
4325     1982.0        13.81                  18.0                0.0469   
4326     1952.0        13.38                  18.0                0.0586   
4327     1924.0        12.51                  18.0                0.0273   

      FuelConsumptionAverage                    Target  
0                   0.000000  [0, 1, 0, 0, 1, 0, 0, 1]  
1                   0.000000  [0, 1, 0, 0, 1, 0, 0, 1]  
2                   0.000000  [0, 1, 0, 0, 1, 0, 0, 1]  
3                   0.000000  [0, 1, 0, 0, 1, 0, 0, 1]  
4                   0.000000  [0, 1, 0, 0, 1, 0, 0, 1]  
...                      ...        

In [17]:
from sklearn.utils import shuffle
df_processed = shuffle(df_processed)

In [19]:
import numpy as np
x_train = df_processed.drop(['Target'], axis=1)
y_train = np.array(df_processed['Target'].values.tolist())
print(x_train.shape)
print(y_train.shape)

(24957, 14)
(24957, 8)


In [20]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x_train, y_train, test_size=0.2, random_state=1)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(19965, 14) (4992, 14) (19965, 8) (4992, 8)


In [24]:
from sklearn.preprocessing import MinMaxScaler

min_max_scaler = MinMaxScaler()
x_scaled = min_max_scaler.fit(X_train)

x_train_transformed = min_max_scaler.transform(X_train)
x_test_transformed = min_max_scaler.transform(X_test)


print(x_train_transformed.shape)
print(x_test_transformed.shape)

(19965, 14)
(4992, 14)


In [25]:
print(y_train.shape)
print(y_test.shape)

(19965, 8)
(4992, 8)


In [26]:
from sklearn.multioutput import MultiOutputClassifier
from sklearn.ensemble import RandomForestClassifier

forest = RandomForestClassifier(random_state=1)
multi_target_forest = MultiOutputClassifier(forest, n_jobs=-1)
multi_target_forest.fit(x_train_transformed, y_train)

MultiOutputClassifier(estimator=RandomForestClassifier(random_state=1),
                      n_jobs=-1)

In [27]:
y_test_predicted = multi_target_forest.predict(x_test_transformed)
y_test_predicted

array([[0, 1, 0, ..., 0, 0, 1],
       [0, 1, 0, ..., 0, 0, 1],
       [0, 0, 1, ..., 1, 0, 1],
       ...,
       [0, 0, 1, ..., 1, 0, 1],
       [1, 0, 0, ..., 0, 0, 1],
       [0, 1, 0, ..., 0, 0, 1]])

In [28]:
print(y_test.shape)
print(y_test_predicted.shape)

(4992, 8)
(4992, 8)


In [29]:
from sklearn.metrics import accuracy_score
score = accuracy_score(y_test, y_test_predicted)
score

0.9128605769230769

In [31]:
from sklearn.tree import DecisionTreeClassifier

forest = DecisionTreeClassifier(random_state=1)
multi_target_forest = MultiOutputClassifier(forest, n_jobs=-1)
multi_target_forest.fit(x_train_transformed, y_train)

y_test_predicted = multi_target_forest.predict(x_test_transformed)
score = accuracy_score(y_test, y_test_predicted)
print(score)

0.8385416666666666


In [32]:
cols = list(one_hot_encoding.columns)[:-1]

ordered_values =[]
for i in cols:
    ordered_values.append(i.split("_")[1])
    
print(ordered_values)
print(len(ordered_values))

['FullOfHolesCondition', 'SmoothCondition', 'UnevenCondition', 'HighCongestionCondition', 'LowCongestionCondition', 'NormalCongestionCondition', 'AggressiveStyle', 'EvenPaceStyle']
8


In [33]:
print(y_test_predicted.shape)

(4992, 8)


In [34]:
def convert_results(y_test_predicted, ordered_values):
    final_prediction = []
    for each_prediction in y_test_predicted:
        sub_list = []
        for i, j in enumerate(list(each_prediction)):
            if j == 1:
                sub_list.append(ordered_values[i])
        final_prediction.append(sub_list)
    return final_prediction


final_prediction = convert_results(y_test_predicted, ordered_values)
final_prediction[:10]

[['SmoothCondition',
  'UnevenCondition',
  'LowCongestionCondition',
  'NormalCongestionCondition',
  'EvenPaceStyle'],
 ['SmoothCondition', 'LowCongestionCondition', 'EvenPaceStyle'],
 ['UnevenCondition', 'NormalCongestionCondition', 'EvenPaceStyle'],
 ['SmoothCondition', 'LowCongestionCondition', 'EvenPaceStyle'],
 ['UnevenCondition', 'HighCongestionCondition', 'EvenPaceStyle'],
 ['FullOfHolesCondition', 'LowCongestionCondition', 'AggressiveStyle'],
 ['UnevenCondition', 'NormalCongestionCondition', 'EvenPaceStyle'],
 ['SmoothCondition', 'NormalCongestionCondition', 'EvenPaceStyle'],
 ['SmoothCondition', 'LowCongestionCondition', 'EvenPaceStyle'],
 ['SmoothCondition', 'LowCongestionCondition', 'AggressiveStyle']]

# Accuracy score for RandomForest Classifier is 92% whereas for Decision Tree, it is 84%. Therefore, RandomForest Classifier will be good choice for classification.